File: midolman/src/main/java/org/midonet/sdn/state/ShardedFlowStateTable.java
package org.midonet.sdn.state;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import com.yammer.metrics.core.Clock;

/**
 * A sharded per-flow state table.
 *
 * It aggregates the data contain in a number of children shards, its goal is
 * to allow for single-writer / multiple-reader synchronization semantics across
 * the aggregated table, by assigning ownership of different shards to different
 * threads.
 *
 * THREADING SEMANTICS:
 *
 * Puts on a shard do not touch other shards. So two writes on the same key
 * performed on different shards result in undefined behaviour. Clients should
 * distribute shards among threads in a way that results in no keyspace overlap.
 *
 * Gets, on the other hand, will fall back to the parent and the other shards
 * if a key is not found locally.
 *
 * unref() calls may require coordination, but they are meant to happen in an
 * external thread or pool, not a shard-owning thread.
 */
public class ShardedFlowStateTable<K, V> implements FlowStateLifecycle<K, V> {
    private ArrayList<FlowStateShard> shards = new ArrayList<>();

    private final int SHARD_NONE = -1;
    Clock clock;

    public ShardedFlowStateTable() {
        this.clock = Clock.defaultClock();
    }

    public ShardedFlowStateTable(Clock clock) {
        this.clock = clock;
    }

    public FlowStateTable<K, V> addShard() {
        FlowStateShard s = new FlowStateShard(shards.size());
        shards.add(s);
        return s;
    }

    /**
     * Fetches a the value associated with a key, skipping the given shard
     * index.
     */
    V get(K key, int shardToSkip) {
        assert(shardToSkip == SHARD_NONE ||
               (shardToSkip >= 0 && shardToSkip < shards.size()));

        for (int i = 0; i < shards.size(); i++) {
            if (i != shardToSkip) {
                V v = shards.get(i).shallowGet(key);
                if (v != null)
                    return v;
            }
        }
        return null;
    }

    @Override
    public V putAndRef(K key, V value) {
        // NO-OP
        throw new IllegalArgumentException();
    }

    @Override
    public V remove(K key) {
        // NO-OP
        throw new IllegalArgumentException();
    }

    @Override
    public V get(K key) {
        return get(key, SHARD_NONE);
    }

    @Override
    public V ref(K key) {
        for (int i = 0; i < shards.size(); i++) {
            V v = shards.get(i).ref(key);
            if (v != null)
                return v;
        }
        return null;
    }

    @Override
    public void unref(K key) {
        for (int i = 0; i < shards.size(); i++) {
            V v = shards.get(i).get(key);
            if (v != null)
                shards.get(i).unref(key);
        }
    }

    @Override
    public <U> U fold(U seed, Reducer<K, V, U> func) {
        for (int i = 0; i < shards.size(); i++) {
            seed = shards.get(i).fold(seed, func);
        }
        return seed;
    }

    @Override
    public <U> U expireIdleEntries(int idleAgeMillis, U seed, Reducer<K, V, U> func) {
        for (int i = 0; i < shards.size(); i++) {
            seed = shards.get(i).expireIdleEntries(idleAgeMillis, seed, func);
        }
        return seed;
    }

    @Override
    public void expireIdleEntries(int idleAgeMillis) {
        for (int i = 0; i < shards.size(); i++) {
            shards.get(i).expireIdleEntries(idleAgeMillis);
        }
    }

    /**
     * A shard within a ShardedFlowStateTable.
     *
     * It stores entries locally but forwards queries to the parent table for
     * aggregation. Reference counting is also delegated on the parent.
     */
    class FlowStateShard implements FlowStateLifecycle<K, V> {
        public static final int IDLE_QUEUE_SIZE = 10000;

        private final int workerId;

        private ConcurrentHashMap<K, Entry> data = new ConcurrentHashMap<>();

        /* NOTE(guillermo): At least for now we will synchronize operations on
         * idleKeys. unref() and idle entry expiration both need to modify it
         * it's not clear that both operations will happen in the same thread. */
        private final PriorityQueue<Entry> idleKeys =
                new PriorityQueue<>(IDLE_QUEUE_SIZE, new EntryComparator());

        public FlowStateShard(int workerId) {
            this.workerId = workerId;
        }

        @Override
        public V putAndRef(K key, V value) {
            V oldV = get(key);
            // TODO(guillermo) reuse Entry instances.
            data.put(key, new Entry(key, value));
            return oldV;
        }

        @Override
        public V remove(K key) {
            Entry e = data.remove(key);
            return e != null ? e.v : null;
        }

        @Override
        public V get(K key) {
            Entry e = data.get(key);
            return e != null ? e.v : ShardedFlowStateTable.this.get(key, workerId);
        }

        public V shallowGet(K key) {
            Entry e = data.get(key);
            return e != null ? e.v : null;
        }

        @Override
        public V ref(K key) {
            Entry e = data.get(key);
            if (e != null) {
                e.ref();
                return e.v;
            }
            return null;
        }

        @Override
        public void unref(K key) {
            Entry e = data.get(key);
            if (e != null && e.unref() == 0) {
                /* We don't mess with 'idleSince' while an entry is in the
                   priority queue */
                synchronized (idleKeys) {
                    if (e.idleSince != 0)
                        idleKeys.remove(e);
                    e.idleSince = clock.tick();
                    idleKeys.add(e);
                }
            }
        }

        @Override
        public <U> U fold(U seed, Reducer<K, V, U> func) {
            for (Entry e: data.values()) {
                seed = func.apply(seed, e.k, e.v);
            }
            return seed;
        }

        @Override
        public void expireIdleEntries(int idleAgeMillis) {
            expireIdleEntries(idleAgeMillis, null, null);
        }

        @Override
        public <U> U expireIdleEntries(int idleAgeMillis, U seed, Reducer<K, V, U> func) {
            long epoch = clock.tick() -
                            TimeUnit.MILLISECONDS.toNanos(idleAgeMillis);
            while (!idleKeys.isEmpty() && idleKeys.peek().idleSince <= epoch) {
                synchronized (idleKeys) {
                    Entry e = idleKeys.remove();
                    if (e.refCount.get() == 0 && data.remove(e.k) != null) {
                        if (func != null)
                            seed = func.apply(seed, e.k, e.v);
                    }
                }
            }
            return seed;
        }

        class Entry {
            K k; V v;
            AtomicInteger refCount;
            long idleSince = 0;

            Entry(K k, V v) {
                this.k = k;
                this.v = v;
                refCount = new AtomicInteger(1);
            }

            int ref() { return refCount.incrementAndGet(); }

            int unref() {
                int n = refCount.decrementAndGet();
                return n >= 0 ? n : refCount.incrementAndGet();
            }
        }

        class EntryComparator implements Comparator<Entry> {
            @Override
            public int compare(Entry a, Entry b) {
                return (int) (a.idleSince - b.idleSince);
            }
        }
    }
}

File: midolman/src/main/scala/org/midonet/sdn/state/ShardedFlowStateTable.scala
/*
 * Copyright (c) 2014 Midokura SARL, All Rights Reserved.
 */

package org.midonet.sdn.state

import java.util.ArrayList

import com.typesafe.scalalogging.Logger
import com.yammer.metrics.core.Clock

import org.midonet.util.concurrent.TimedExpirationMap
import org.midonet.util.collection.Reducer
import org.slf4j.LoggerFactory

object ShardedFlowStateTable {
    def create[K <: IdleExpiration, V >: Null](): ShardedFlowStateTable[K, V] =
            new ShardedFlowStateTable[K, V]()

    def create[K <: IdleExpiration, V >: Null](clock: Clock):
            ShardedFlowStateTable[K, V] = new ShardedFlowStateTable[K, V](clock)
}

/**
 * A sharded per-flow state table.
 *
 * It aggregates the data contain in a number of children shards, its goal is
 * to allow for single-writer / multiple-reader synchronization semantics across
 * the aggregated table, by assigning ownership of different shards to different
 * threads.
 *
 * THREADING SEMANTICS:
 *
 * Puts on a shard do not touch other shards. So two writes on the same key
 * performed on different shards result in undefined behaviour. Clients should
 * distribute shards among threads in a way that results in no keyspace overlap.
 *
 * Gets, on the other hand, will fall back to the parent and the other shards
 * if a key is not found locally.
 *
 * unref() calls may require coordination, but they are meant to happen in an
 * external thread or pool, not a shard-owning thread.
 */
class ShardedFlowStateTable[K <: IdleExpiration, V >: Null]
        (val clock: Clock = Clock.defaultClock()) extends FlowStateTable[K, V] {

    private val shards = new ArrayList[FlowStateShard]()
    private val SHARD_NONE: Int = -1

    private val defaultLogger =
        Logger(LoggerFactory.getLogger("org.midonet.state.table"))

    def addShard(log: Logger = defaultLogger) = {
        val s: FlowStateShard = new FlowStateShard(shards.size, log)
        shards.add(s)
        s
    }

    /**
     * Fetches a the value associated with a key, skipping the given shard
     * index.
     */
    private[state] def get(key: K, shardToSkip: Int): V = {
        assert(shardToSkip == SHARD_NONE ||
                (shardToSkip >= 0 && shardToSkip < shards.size))

        var i = 0
        while (i < shards.size) {
            if (i != shardToSkip) {
                val v: V = shards.get(i).shallowGet(key)
                if (v != null)
                    return v
            }
            i += 1
        }
        null
    }

    override def putAndRef(key: K, value: V): V = throw new IllegalArgumentException

    override def get(key: K): V = get(key, SHARD_NONE)

    override def ref(key: K): V = {
        var i = 0
        while (i < shards.size) {
            val v = shards.get(i).ref(key)
            if (v != null)
                return v
            i += 1
        }
        null
    }

    override def touch(key: K, value: V) {
        var i: Int = 0
        while (i < shards.size) {
            shards.get(i).touch(key, value)
            i += 1
        }
    }

    override def getRefCount(key: K): Int = {
        var count = 0
        var i = 0
        while (i < shards.size) {
            count += shards.get(i).getRefCount(key)
            i += 1
        }
        count
    }

    override def unref(key: K) {
        var i: Int = 0
        while (i < shards.size) {
            val v: V = shards.get(i).shallowGet(key)
            if (v != null)
                shards.get(i).unref(key)
            i += 1
        }
    }

    override def fold[U](acc: U, func: Reducer[K, V, U]): U = {
        var i = 0
        var seed = acc
        while (i < shards.size) {
            seed = shards.get(i).fold(seed, func)
            i += 1
        }
        seed
    }

    override def expireIdleEntries[U](acc: U, func: Reducer[K, V, U]): U = {
        var i = 0
        var seed = acc
        while (i < shards.size) {
            seed = shards.get(i).expireIdleEntries(seed, func)
            i += 1
        }
        seed
    }

    override def expireIdleEntries() {
        var i = 0
        while (i < shards.size) {
            shards.get(i).expireIdleEntries()
            i += 1
        }
    }

    /**
     * A shard within a ShardedFlowStateTable.
     *
     * It stores entries locally but forwards queries to the parent table for
     * aggregation. Reference counting is also delegated on the parent.
     */
    class FlowStateShard(workerId: Int, log: Logger) extends FlowStateTable[K, V] {
        private val map = new TimedExpirationMap[K, V](log, _.expiresAfter)

        override def putAndRef(key: K, value: V): V =
            map.putAndRef(key, value)

        override def get(key: K) = {
            val v = map.get(key)
            if (v != null)
                v
            else
                ShardedFlowStateTable.this.get(key, workerId)
        }

        def shallowGet(key: K): V =
            map.get(key)

        override def ref(key: K): V =
            map.ref(key)

        override def getRefCount(key: K): Int =
            map.getRefCount(key)

        override def touch(key: K, value: V): Unit = {
            putAndRef(key, value)
            unref(key)
        }

        private def tickMillis = clock.tick / 1000000

        override def unref(key: K) =
            map.unref(key, tickMillis)

        override def fold[U](seed: U, func: Reducer[K, V, U]): U =
            map.fold(seed, func)

        override def expireIdleEntries() =
            map.obliterateIdleEntries(tickMillis)

        override def expireIdleEntries[U](seed: U, func: Reducer[K, V, U]): U =
            map.obliterateIdleEntries(tickMillis, seed, func)
    }
}

